src/Plugin/GraphQL/DataProducer/EntityValidationTrait.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace Drupal\graphql\Plugin\GraphQL\DataProducer;
+
+use Drupal\Core\Entity\ContentEntityInterface;
+
+/**
+ * Trait for entity validation.
+ *
+ * Ensure the entity passes validation, any violations will be reported back
+ * to the client. Validation will catch issues like invalid referenced entities,
+ * incorrect text formats, required fields etc. Additional validation of input
+ * should not be put here, but instead should be built into the entity
+ * validation system, so the same constraints are applied in the Drupal admin.
+ */
+trait EntityValidationTrait {
+
+  /**
+   * Get violation messages from an entity.
+   *
+   * @param \Drupal\Core\Entity\ContentEntityInterface $entity
+   *   An entity to validate.
+   *
+   * @return array|null
+   *   Get a list of violations or NULL if none were found.
+   */
+  public function getViolationMessages(ContentEntityInterface $entity) {
+    $violations = $entity->validate();
```

Comment:
we need to filter out validation errors that the user does not have access to, like EntityValidationTrait in the JSONAPI core module.

Reply:
Done, but no test added just yet.